crates/corro-agent/src/agent/handlers.rs

```diff
@@ -0,0 +1,505 @@
+                    }
+                }
+                increment_counter!("corro.swim.notification", "type" => "memberup");
+            }
+            Notification::MemberDown(actor) => {
+                let removed = { agent.members().write().remove_member(&actor) };
+                trace!("Member Down {actor:?} (removed: {removed})");
+                if removed {
+                    debug!("Member Down {actor:?}");
+                    increment_counter!("corro.gossip.member.removed", "id" => actor.id().0.to_string(), "addr" => actor.addr().to_string());
+                    // actually removed a member
+                    // notify of new cluster size
+                    let member_len = { agent.members().read().states.len() as u32 };
+                    if let Ok(size) = member_len.try_into() {
+                        if let Err(e) = agent.tx_foca().send(FocaInput::ClusterSize(size)).await {
+                            error!("could not send new foca cluster size: {e}");
+                        }
+                    }
+                }
+                increment_counter!("corro.swim.notification", "type" => "memberdown");
+            }
+            Notification::Active => {
+                info!("Current node is considered ACTIVE");
+                increment_counter!("corro.swim.notification", "type" => "active");
+            }
+            Notification::Idle => {
+                warn!("Current node is considered IDLE");
+                increment_counter!("corro.swim.notification", "type" => "idle");
+            }
+            // this happens when we leave the cluster
+            Notification::Defunct => {
+                debug!("Current node is considered DEFUNCT");
+                increment_counter!("corro.swim.notification", "type" => "defunct");
+            }
+            Notification::Rejoin(id) => {
+                info!("Rejoined the cluster with id: {id:?}");
+                increment_counter!("corro.swim.notification", "type" => "rejoin");
+            }
+        }
+    }
+}
+
+// DOCME: provide function context
+pub async fn handle_db_cleanup(pool: SplitPool) -> eyre::Result<()> {
+    debug!("handling db_cleanup (WAL truncation)");
+    let conn = pool.write_low().await?;
+    block_in_place(move || {
+        let start = Instant::now();
+
+        let busy: bool =
+            conn.query_row("PRAGMA wal_checkpoint(TRUNCATE);", [], |row| row.get(0))?;
+        if busy {
+            warn!("could not truncate sqlite WAL, database busy");
+            increment_counter!("corro.db.wal.truncate.busy");
+        } else {
+            debug!("successfully truncated sqlite WAL!");
+            histogram!(
+                "corro.db.wal.truncate.seconds",
+                start.elapsed().as_secs_f64()
+            );
+        }
+        Ok::<_, eyre::Report>(())
+    })?;
+    debug!("done handling db_cleanup");
+    Ok(())
+}
+
+// DOCME: provide function context
```

Comment:
This is the function that bundles changes to optimize transactions sizes with SQLite which makes everything more efficient at the cost of a little bit of waiting for more data before processing it.

It's used by both broadcast receivers and sync receivers so it's all one happy family.